### PR TITLE
Add Save, SaveAll, Restore, RestoreAll to pkg/util/iptables

### PR DESCRIPTION
**Pull request by BenTheElder**
Adds utility wrappers for `iptables-save` and `iptables-restore` to the
iptables Interface in pkg/util/iptables. Also const’s command strings.

### CODE REVIEW

**Comment by BenTheElder** [comment]
This is part of work for implementing https://github.com/GoogleCloudPlatform/kubernetes/issues/3760; see larger PR: https://github.com/GoogleCloudPlatform/kubernetes/pull/9210
/cc @thockin

File: pkg/util/iptables/iptables.go [hunk 1]
```diff
@@ -178,11 +211,112 @@ func (runner *runner) IsIpv6() bool {
...
+	if !flush {
+		args = append(args, "--noflush")
+	}
+	if counters {
+		args = append(args, "--counters")
+	}
+	// create temp file through which to pass data
+	temp, err := ioutil.TempFile("", "kube-temp-iptables-restore-")
+	// make sure we delete the temp file
+	defer os.Remove(temp.Name())
```
Comment (thockin) [inline pkg/util/iptables/iptables.go]: only defer if err == nil

File: pkg/util/iptables/iptables.go [hunk 2]
```diff
@@ -178,11 +211,112 @@ func (runner *runner) IsIpv6() bool {
...
```
Comment (thockin) [inline pkg/util/iptables/iptables.go]: same comment

File: pkg/util/iptables/iptables.go [hunk 3]
```diff
@@ -178,11 +211,112 @@ func (runner *runner) IsIpv6() bool {
...
+}
+
+// Restore is part of Interface.
+func (runner *runner) Restore(table Table, data []byte, flush FlushFlag, counters RestoreCountersFlag) error {
+	runner.mu.Lock()
+	defer runner.mu.Unlock()
+
+	// setup args
+	args := []string{"-T", string(table)}
+	if !flush {
```
Comment (thockin) [inline pkg/util/iptables/iptables.go]: can you capture lines 240-271 as a private method and call it from both Restore() and RestoreAll() ?

File: pkg/util/iptables/iptables.go (hunk 1 above)
Comment (BenTheElder) [inline pkg/util/iptables/iptables.go]: fixed.

File: pkg/util/iptables/iptables.go [hunk 4]
```diff
@@ -49,6 +51,18 @@ type Interface interface {
 	DeleteRule(table Table, chain Chain, args ...string) error
 	// IsIpv6 returns true if this is managing ipv6 tables
 	IsIpv6() bool
+	// Save calls `iptables-save` for table.
+	Save(table Table) ([]byte, error)
```
Comment (thockin) [inline pkg/util/iptables/iptables.go]: Does the addition of these methods break any other call-sites that assume they are implementing iptables.Interface ?

Comment (thockin) [inline pkg/util/iptables/iptables.go]: argument to createProxier:
*fakeIptables does not implement iptables.Interface (missing Restore method)
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/proxier_test.go:508: cannot use fakeIptables literal (type *fakeIptables) as type iptables.Interface in argument to createProxier:
*fakeIptables does not implement iptables.Interface (missing Restore method)
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/proxier_test.go:560: cannot use fakeIptables literal (type *fakeIptables) as type iptables.Interface in argument to createProxier:
*fakeIptables does not implement iptables.Interface (missing Restore method)
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/proxier_test.go:560: too many errors
FAIL	k8s.io/kubernetes/pkg/proxy [build failed]

File: pkg/util/iptables/iptables.go [hunk 5]
```diff
@@ -49,6 +51,18 @@ type Interface interface {
...
+	// SaveAll calls `iptables-save`.
+	SaveAll() ([]byte, error)
```
Comment (thockin) [inline pkg/util/iptables/iptables.go]: These really should have unit tests.  I will accept a TODO as long as you promise to do them :)

The tests won't be hard, but they will ensure that what happens is what we expect.

File: pkg/util/iptables/iptables.go (hunk 4 above)
Comment (BenTheElder) [inline pkg/util/iptables/iptables.go]: Ah yes, one callsite that mocks it in the userspace proxier tests, will
patch that in shortly. Forgot that, quick-release built fine, tests break
from this.

File: pkg/util/iptables/iptables.go (hunk 3 above)
Comment (BenTheElder) [inline pkg/util/iptables/iptables.go]: yes

File: pkg/util/iptables/iptables.go (hunk 4 above)
Comment (BenTheElder) [inline pkg/util/iptables/iptables.go]: the code for this exists in the iptables proxy branch, just need to pull over here.

File: pkg/util/iptables/iptables.go (hunk 3 above)
Comment (thockin) [inline pkg/util/iptables/iptables.go]: let me try again, I know you CAN, but would you please do so :)

Comment (BenTheElder) [inline pkg/util/iptables/iptables.go]: Understood first time, silly reply on my part just checking off each comment as I fix up the next commit.

Comment (thockin) [inline pkg/util/iptables/iptables.go]: I was being a smart ass - I have a 5 year old who would say yes and stare at me. :)

**Comment by thockin** [comment]
LGTM
